File: Java/69-Sqrt(x).java
public class Solution {
    public int mySqrt(int x) {
        if (x <= 0){
            return 0;
        }
        if (x == 1){
            return 1;
        }
        int left = 1, right = x;
        while (left < right){
            int mid = left + (right - left) / 2;
            if (mid < x / mid){
                left = mid + 1;
            }else if (mid == x / mid){
                return mid;
            }else{
                right = mid;
            }
        }
        return left - 1;
    }
}

File: cpp/69-Sqrt(x).cpp
class Solution {
public:
    int mySqrt(int x) {
        int start = 0, end = x;
        while (start <= end) {
            long mid = start + (end - start) / 2;
            if (mid * mid == x) {
                return mid;
            }else if (mid * mid > x) {
                end = mid - 1;
            }else {
                start = mid + 1;
            }
        }
        //because int rounds down.
        return end;
    }
};
